modules/accounts/src/lib.rs

```diff
@@ -62,29 +77,91 @@ decl_module! {
 			let who = ensure_signed(origin)?;
 
-			ensure!(T::DepositCurrency::free_balance(&who) > T::FreeTransferDeposit::get(), Error::<T>::NotEnoughBalance);
+			let native_currency_id = T::NativeCurrencyId::get();
+			let free_transfer_deposit = T::FreeTransferDeposit::get();
+			ensure!(<T as Trait>::Currency::free_balance(native_currency_id, &who) > free_transfer_deposit, Error::<T>::NotEnoughBalance);
 
-			T::DepositCurrency::set_lock(ACCOUNTS_ID, &who, T::FreeTransferDeposit::get(), WithdrawReasons::all());
+			<T as Trait>::Currency::set_lock(ACCOUNTS_ID, native_currency_id, &who, T::FreeTransferDeposit::get());
 			<FreeTransferEnabledAccounts<T>>::insert(who, true);
 		}
 
 		#[weight = 10_000]
 		fn disable_free_transfers(origin) {
 			let who = ensure_signed(origin)?;
 
-			T::DepositCurrency::remove_lock(ACCOUNTS_ID, &who);
+			<T as Trait>::Currency::remove_lock(ACCOUNTS_ID, T::NativeCurrencyId::get(), &who);
 			<FreeTransferEnabledAccounts<T>>::remove(who);
 		}
+
+		#[weight = 0]
+		fn close_account(origin, to: Option<T::AccountId>) {
+			let who = ensure_signed(origin)?;
+
+			// check must allow death
+			// if native/non-native currencies has locks, means ref_count shouldn't be zero, can not close the account
+			ensure!(
+				<system::Module<T>>::allow_death(&who),
+				Error::<T>::NonZeroRefCount,
+			);
+
+			let native_currency_id = T::NativeCurrencyId::get();
+			let treasury_account = Self::treasury_account_id();
+			let specific_receiver = to.unwrap_or_else(|| treasury_account.clone());
+
+			let total_reserved_native = <T as Trait>::Currency::reserved_balance(native_currency_id, &who);
+
+			// unreserve all native currency
+			<T as Trait>::Currency::unreserve(native_currency_id, &who, total_reserved_native);
+
+			// reserved except for the NewAccountDeposit should be refund to `TreasuryModuleId`
+			if let Some(refund_to_treasury_reserved) = total_reserved_native.checked_sub(T::NewAccountDeposit::get()) {
+				// transfer refund to treasury seperately if specific_receiver is not treasury_account
+				if treasury_account != specific_receiver {
+					<T as Trait>::Currency::transfer(native_currency_id, &who, &treasury_account, refund_to_treasury_reserved)?;
+				}
+			}
+
+			// transfer all free to specific_receiver
+			<T as Trait>::Currency::transfer(native_currency_id, &who, &specific_receiver, <T as Trait>::Currency::free_balance(native_currency_id, &who))?;
+
+			// handle other non-native currencies
+			for currency_id in T::AllCurrencyIds::get() {
+				let reserved = <T as Trait>::Currency::reserved_balance(currency_id, &who);
+				if !reserved.is_zero() {
+					// unreserve all reserved
+					<T as Trait>::Currency::unreserve(currency_id, &who, reserved);
+
+					// transfer reserved amount to treasury_account seperately if specific_receiver is not treasury_account
+					if treasury_account != specific_receiver {
+						<T as Trait>::Currency::transfer(currency_id, &who, &treasury_account, reserved)?;
```

Comment:
```suggestion
						let _ = <T as Trait>::Currency::transfer(currency_id, &who, &treasury_account, reserved);
```

should just ignore error